src/gui/transferlistmodel.cpp

```diff
@@ -179,37 +203,175 @@ QVariant TransferListModel::headerData(int section, Qt::Orientation orientation,
     return {};
 }
 
-QVariant TransferListModel::data(const QModelIndex &index, const int role) const
+QString TransferListModel::displayValue(const BitTorrent::TorrentHandle *torrent, int column) const
 {
-    if (!index.isValid()) return {};
-
-    const BitTorrent::TorrentHandle *torrent = m_torrentList.value(index.row());
-    if (!torrent) return {};
-
-    if ((role == Qt::DecorationRole) && (index.column() == TR_NAME))
-        return getIconByState(torrent->state());
-
-    if (role == Qt::ForegroundRole)
-        return stateForeground(torrent->state());
+    const bool isHideState = (Preferences::instance()->getHideZeroComboValues() == 1)
+            && (torrent->state() == BitTorrent::TorrentState::PausedDownloading); // paused torrents only
+    const bool hideValues = Preferences::instance()->getHideZeroValues() && isHideState;
+
+    const auto availabilityString = [hideValues](const qreal value)
```

Comment:
Return type deduction confuses you generally or in some special cases?

Reply:
No, you write it out so compiler could check for you if you returned the correct type and not something else.

Reply:
I trust myself in such trivial cases.
But I will change this to make others feel safer "under compiler protection". The same goes for your comments about "const" in function parameters.

Reply:
>I trust myself in such trivial cases.

Even if that is true, this is a public project and others might trip over.

>But I will change this to make others feel safer "under compiler protection". The same goes for your comments about "const" in function parameters.

Thank you your majesty.

Reply:
BTW, in case of lambda return type compiler will say you about something wrong anyway (e.g. in place where you call it).
>Even if that is true, this is a public project and others might trip over.

Just for fun. How many times have we (qBittorrent project) encountered, for example, this terrible problem of "unintentional variable change" (before you start insisting on `const` everywhere)?

Reply:
>BTW, in case of lambda return type compiler will say you about something wrong anyway (e.g. in place where you call it).

Yes, but that error message won't be pointing the fault at the lambda which is less correct.

>How many times have we (qBittorrent project) encountered, for example, this terrible problem of "unintentional variable change

Can't be sure, however when a variable is marked const it is greatly easier to track its state without worrying it has changed over a bunch of code. And BTW, in the newer Rust language, every variable is const by default unless you explicitly add `mut` to it.

Reply:
>in Rust language, every variable is const by default unless you explicitly add mut to it.

It's good.
But in C++ you're making me do extra low-effeciency work (add a bunch of "const" modifiers).
>it has changed over a bunch of code.

But in many cases it's just a couple of lines (passing to another function or return statement) so I don't feel it's so serious that I can develop a habit to blindly add const everywhere. You'll have to remind me of that.

Reply:
>Yes, but that error message won't be pointing the fault at the lambda which is less correct.

Lambda can be either correct or no. In first case compiler points to error in place of its incorrect use, in the second one compiler points to error in lambda. I don't understand what's the problem you worry about.

Reply:
>But in C++ you're making me do extra low-effeciency work (add a bunch of "const" modifiers).

At the cost of you spent at most 2 more mins and ease other developers mind set reading your code... and you think it isn't worth it. Funny.

>You'll have to remind me of that.

Don't worry, I will.

>In first case compiler points to error in place of its incorrect use, in the second one compiler points to error in lambda. I don't understand what's the problem you worry about.

The same as you would write `auto someFunc() {}`. I think you should start writing in that way (automatic return type deduction since c++14). /s
But seriously I can't understand why you would expect the lambdas to return other types than `QString` here and also why merely adding `QString` would lead to this much resistance?

If you really think your code is just superior for everyone, then just say it! I'll blindly press the Approve button for all of your PR from now on. @sledgehammer999

Reply:
>If you really think your code is just superior for everyone, then just say it!

No, that's why I'm discussing it. But if you take my arguments as a personal insult (If you really think YOUR code is just superior for everyone), I will not continue this discussion and simply apply your recommendations.
>I think I'll do selective review for project's PR from now on. I.e. no need to wait for me for PRs that I didn't comment.

Well, that will certainly help the project. Many PRs have every chance of never being merged if they have to wait for @sledgehammer999 review. If this applies to my PRs only, then it will be easier for me to stop them altogether (hopefully this is not what you're after).

When I start discussing something, I want to learn something new for myself or correct my misconception (and do the same for an opponent). But it was never my purpose to offend anyone. So let's keep working on the project the way it's best for it.

Reply:
@Chocobo1, PR updated. I hope I have corrected all my "mistakes".

Reply:
I understand that you already pushed the change to the lambda. But shouldn't this be better addressed by **not** using `auto` in this case? The compiler would complain if the lambda didn't return a `QString` object **and** the reader would know what the type of the variable is.

Reply:
Ignore my previous comment. But, IMO, it is too much to require to explicitly declare a return type for such small function.
And any possible mismatch will be realized the moment we try to assign the return type to a 'QString` and the compiler throws an error.

Reply:
>But, IMO, it is too much to require to explicitly declare a return type for such small function.

But you write out the exact return type for normal functions instead of plugging `auto` in there and why is that?

>And any possible mismatch will be realized the moment we try to assign the return type to a 'QString` and the compiler throws an error.

And then trace the error back to the faulty lambda, why not just specify the type at lambda instead and compiler will point to the exact line that cause the error? (I'm talking about general cases)
I may be a little inconsistent on this: for POD types I might not feel strongly the need to write it out, but for more complex types such as Qt containers I think it is a must.

Reply:
>But you write out the exact return type for normal functions instead of plugging auto in there and why is that?

Actually you may be right. For me it is about readability.
So yeah, in lambas the compiler might be able to deduce types and eventually error out, but when a developer reads the lambda he should be able to know the return type from the signature, and not mentally parse the lambda body to understand what type is returned.

Almost same reason why we use `auto` [sparingly](https://github.com/qbittorrent/qBittorrent/blob/master/CODING_GUIDELINES.md#9-misc) in variable types

Reply:
>But you write out the exact return type for normal functions instead of plugging auto in there and why is that?

It is incorrect to put lambda in direct correspondence with the regular function.
Often the function signature is the only thing available (at least easily available). Its implementation can be in another file, library, or available only in binary form.
Lambda was created primarily for "inplace" use, so that excessive detail here only reduces readability. If it is quite complex/large, then this is a good reason to turn it into regular function.

>when a developer reads the lambda he should be able to know the return type from the signature, and not mentally parse the lambda body to understand what type is returned.
Almost same reason why we use auto sparingly in variable types

Personally I don't see any problems with identifying the return type in the lambda like the following:
```c++
[](int param1, bool param2)
{
    return param2 ? QString {} : QString::number(param1);
}
```
IMO, it's better than the following "readable" signature:
```c++
[](const int param1, const bool param2) -> QString
{
    return param2 ? QString {} : QString::number(param1);
}
```